miden-crypto/src/dsa/ecdsa_secp256k1/mod.rs

```diff
@@ -0,0 +1,397 @@
+    ecdh::{EphemeralPublicKey, SharedSecret},
+    utils::{ByteReader, ByteWriter, Deserializable, DeserializationError, Serializable},
+};
+
+#[cfg(test)]
+mod tests;
+
+// CONSTANTS
+// ================================================================================================
+
+/// Length of secret key in bytes
+const SECRET_KEY_BYTES: usize = 32;
+/// Length of public key in bytes when using compressed format encoding
+pub(crate) const PUBLIC_KEY_BYTES: usize = 33;
+/// Length of signature in bytes using our custom serialization
+const SIGNATURE_BYTES: usize = 66;
+/// Length of signature in bytes using standard serialization i.e., `SEC1`
+const SIGNATURE_STANDARD_BYTES: usize = 64;
+/// Length of scalars for the `secp256k1` curve
+const SCALARS_SIZE_BYTES: usize = 32;
+
+// ECDSA HASHER
+// ================================================================================================
+
+#[repr(u8)]
+#[derive(Debug, Clone, Copy)]
+pub enum EcdsaHasher {
+    Sha256 = 0,
+    Keccak = 1,
+}
+
+impl EcdsaHasher {
+    /// Hashes a message using the variant of the ECDSA hasher.
+    fn hash(&self, message: &[u8]) -> [u8; 32] {
+        match self {
+            EcdsaHasher::Sha256 => {
+                use k256::sha2::{Digest, Sha256};
+                let mut hasher = Sha256::new();
+                hasher.update(message);
+                hasher.finalize().into()
+            },
+            EcdsaHasher::Keccak => {
+                use sha3::{Digest, Keccak256};
+                let mut hasher = Keccak256::new();
+                hasher.update(message);
+                hasher.finalize().into()
+            },
+        }
+    }
+
+    /// Converts the hasher variant to a `u8`.
+    pub fn to_byte(&self) -> u8 {
+        *self as u8
+    }
+}
+
+impl TryFrom<u8> for EcdsaHasher {
+    type Error = HasherError;
+
+    fn try_from(byte: u8) -> Result<Self, Self::Error> {
+        match byte {
+            0 => Ok(EcdsaHasher::Sha256),
+            1 => Ok(EcdsaHasher::Keccak),
+            _ => Err(HasherError::InvalidHasher(byte)),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum HasherError {
+    #[error("Invalid hasher byte value: {0}")]
+    InvalidHasher(u8),
+}
+
+// SECRET KEY
+// ================================================================================================
+
+/// Secret key for ECDSA signature verification over secp256k1 curve.
+pub struct SecretKey {
+    inner: SigningKey,
+}
+
+impl SecretKey {
+    /// Generates a new random secret key using the OS random number generator.
+    #[cfg(feature = "std")]
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        let mut rng = rand::rng();
+
+        Self::with_rng(&mut rng)
+    }
+
+    /// Generates a new secret key using the provided random number generator.
+    pub fn with_rng<R: CryptoRng + RngCore>(_rng: &mut R) -> Self {
+        let mut rng = k256::elliptic_curve::rand_core::OsRng;
```

Comment:
Does this ignore the provided `_rng` and create a new `rng` to get randomness from the operating system? If so, would this be `no_std`-compatible?